src/placeos-core/pulse/pulse.cr

```diff
@@ -0,0 +1,96 @@
+require "placeos-models"
+require "ulid"
+require "hashcash"
+# require "placeos"
+require "rethinkdb-orm"
+require "http/client"
+require "tasker"
+require "sodium"
+require "uri"
+
+class PlaceOS::Core::Pulse
+  private getter instance_id : String
+  private getter secret_key : Sodium::Sign::SecretKey
+
+  private getter task : Tasker::Repeat(HTTP::Client::Response)
+
+  def initialize(
+    @instance_id : String,
+    secret_key : String?,
+    heartbeat_interval : Time::Span = 1.day
+  )
+    @secret_key = Sodium::Sign::SecretKey.new(secret_key.hexbytes)
+    @task = Tasker.every(heartbeat_interval) { heartbeat }
+  end
+
+  def heartbeat
+    Message.new(@instance_id, @secret_key).send
+  end
+
+  def finalize
+    @task.cancel
+  end
+
+  def register
+    # from portal pls fix
+    # how a signature is created
+    message = {
+      "systems_qty" => 12,
+      "levels_qty"  => 30,
+    }.to_json
+
+    sec_key_string = "b18e1d0045995ec3d010c387ccfeb984d783af8fbb0f40fa7db126d889f6dadd77f48b59caeda77751ed138b0ec667ff50f8768c25d48309a8f386a2bad187fb"
+    pub_key_string = "77f48b59caeda77751ed138b0ec667ff50f8768c25d48309a8f386a2bad187fb"
+    key = Sodium::Sign::SecretKey.new(sec_key_string.hexbytes, pub_key_string.hexbytes)
+    sig = key.sign_detached(message).hexstring
+
+    json = {
+      "instance_id" => "01EZZRAS54NA896VS67XEVF0G8",
+      "message"     => JSON.parse(message),
+      "signature"   => sig,
+    }.to_json
+    instance_request = PlaceOS::Portal::Api::InstanceRequest.from_json(json)
+  end
+end
+
+class Message < PlaceOS::Core::Pulse
```

Comment:
All of these classes should follow the class hierarchy
i.e `Message` is part of `Pulse`, so it should be beneath `Pulse`, not a top-level constant